File: builder/java/Phone.java
package androidhot.github.io.designpatterndemo.builder.java;

public class Phone {
  // Required parameters
  private final String mName;
  private final float mPrice;

  // Optional parameters
  private final String mCpu;
  private final int mMemory;
  private final int mStorage;
  private final float mWeight;

  private Phone(Builder builder) {
    mName = builder.name;
    mPrice = builder.price;
    mCpu = builder.cpu;
    mMemory = builder.memory;
    mStorage = builder.storage;
    mWeight = builder.weight;
  }

  // some function

  public static class Builder {
    // Required parameters
    private final String name;
    private final float price;

    // Optional parameters, initialized to default values.
    private String cpu = "";
    private int memory = 0;
    private int storage = 0;
    private float weight = 0;

    public Builder(String name, float price) {
      this.name = name;
      this.price = price;
    }

    public Builder setCpu(String cpu) {
      this.cpu = cpu;
      return this;
    }

    public Builder setMemory(int memory) {
      this.memory = memory;
      return this;
    }

    public Builder setStorage(int storage) {
      this.storage = storage;
      return this;
    }

    public Builder setWeight(float weight) {
      this.weight = weight;
      return this;
    }

    public Phone build() {
      return new Phone(this);
    }
  }
}

File: builder/kotlin/Phone.kt
package androidhot.github.io.designpatterndemo.builder.kotlin

class Phone private constructor(builder: Builder) {
    // Required parameters
    private val mName: String
    private val mPrice: Float

    // Optional parameters
    private val mCpu: String
    private val mMemory: Int
    private val mStorage: Int
    private val mWeight: Float

    init {
        mName = builder.getName()
        mPrice = builder.getPrice()
        mCpu = builder.getCpu()
        mMemory = builder.getMemory()
        mStorage = builder.getStorage()
        mWeight = builder.getWeight()
    }

    // some function

    class Builder(private val name: String, private val price: Float) {

        // Optional parameters, initialized to default values.
        private var cpu = ""
        private var memory = 0
        private var storage = 0
        private var weight = 0f

        fun getName(): String {
            return name
        }

        fun getPrice(): Float {
            return price
        }

        fun setCpu(cpu: String): Builder {
            this.cpu = cpu
            return this
        }

        fun getCpu(): String {
            return cpu
        }

        fun setMemory(memory: Int): Builder {
            this.memory = memory
            return this
        }

        fun getMemory(): Int {
            return memory
        }

        fun setStorage(storage: Int): Builder {
            this.storage = storage
            return this
        }

        fun getStorage(): Int {
            return storage
        }

        fun setWeight(weight: Float): Builder {
            this.weight = weight
            return this
        }

        fun getWeight(): Float {
            return weight
        }

        fun build(): Phone {
            return Phone(this)
        }
    }
}
